components/calendar/src/japanese.rs

```diff
@@ -210,17 +236,38 @@ const REIWA_START: EraStartDate = EraStartDate {
 const FALLBACK_ERA: (EraStartDate, TinyStr16) = (REIWA_START, tinystr!(16, "reiwa"));
 
 impl Japanese {
-    /// Given an ISO date, obtain the era data
-    fn era_for(&self, date: &IsoDateInner) -> (EraStartDate, TinyStr16) {
+    /// Given an ISO date, give year and era for that date in the Japanese calendar
+    ///
+    /// This will also use Gregorian eras for eras that are before the earliest era
+    fn adjusted_year_for(&self, date: &IsoDateInner) -> (i32, TinyStr16) {
         let date: EraStartDate = date.into();
+        let (start, era) = self.japanese_era_for(date);
+        // The year in which an era starts is Year 1, and it may be short
+        // The only time this function will experience dates that are *before*
+        // the era start date are for the first era (Currently, taika-645
+        // for japanext, meiji for japanese),
+        // In such a case, we instead fall back to Gregorian era codes
+        if date < start {
+            if date.year < 0 {
+                (1 - date.year, tinystr!(16, "bce"))
+            } else {
+                (date.year, tinystr!(16, "ce"))
+            }
+        } else {
+            (date.year - start.year + 1, era)
+        }
+    }
+
+    /// Given an date, obtain the era data (not counting spliced gregorian eras)
+    fn japanese_era_for(&self, date: EraStartDate) -> (EraStartDate, TinyStr16) {
         let era_data = self.eras.get();
         // We optimize for the five "modern" post-Meiji eras, which are stored in a smaller
         // array and also hardcoded. The hardcoded version is not used if data indicates the
         // presence of newer eras.
         if date >= MEIJI_START {
-            if era_data.dates_to_eras.len() == 5 {
+            if era_data.dates_to_eras.last().map(|x| x.1) == Some(tinystr!(16, "reiwa")) {
```

Comment:
Thought: Feels like premature optimization

Reply:
I'd expect `.year()` to be fast which is why I left this in.